### PR TITLE
Extended canHandleSource in media files to consistently guess mime types. Fixes #1833

### PR DESCRIPTION
**Pull request by carpasse**
Added getFileExtension to libs. And used it on Html5 and Flash canHandleSource method to guess the mime type if necessary.

### CODE REVIEW

**Comment by heff** [comment]
Awesome, thanks!

File: src/js/lib.js [hunk 1]
```diff
@@ -849,3 +849,25 @@ vjs.arr.forEach = function(array, callback, thisArg) {
...
+ *
+ * @param {String}    path    The fileName path like '/path/to/file.mp4'
+ * @returns {String}          The extension in lower case or an empty string if no extension could be found.
+ */
+vjs.getFileExtension = function(path) {
+  var splitPathRe;
+  var pathParts;
+
+  if(typeof path === 'string'){
+    splitPathRe = /^(\/?)([\s\S]*?)((?:\.{1,2}|[^\/]+?)(\.([^\.\/\?]+)))(?:[\/]*|[\?].*)$/i;
```
Comment (heff) [inline src/js/lib.js]: That's a pretty advanced regex. What does it add over the one that was in the html5 tech?

File: src/js/media/flash.js [hunk 2]
```diff
@@ -233,15 +233,24 @@ vjs.Flash.nativeSourceHandler = {};
  * @param  {Object} source  The source object
  * @return {String}         'probably', 'maybe', or '' (empty string)
  */
-vjs.Flash.nativeSourceHandler.canHandleSource = function(source){
+
+vjs.Flash.nativeSourceHandler.canHandleSource = function (source) {
   var type;
 
-  if (!source.type) {
+  function guessMimeType(src) {
```
Comment (heff) [inline src/js/media/flash.js]: Is it worth adding a function here if it's used once? It does scope the var so if you think it's just cleaner that's fine with me.

File: test/unit/lib.js [hunk 3]
```diff
@@ -403,3 +403,22 @@ test('should loop through each element of an array', function() {
     }
   });
 });
+
+//getFileExtension tests
+test('should get the file extension of the passed path', function() {
```
Comment (heff) [inline test/unit/lib.js]: Can we add a test for a full URL? e.g. `http://...` I can see that it supports it but it'd be good to have a test for it.

**Comment by heff** [comment]
Added  a few comments, but otherwise this looks good to me. :+1:

File: src/js/lib.js (hunk 1 above)
Comment (gkatsev) [inline src/js/lib.js]: It's the regex from [iojs](https://github.com/iojs/io.js/blob/4e9bf93e9c645f33446f7d47005868ffcb3d281e/lib/path.js#L394), which we talked about in the [other pr comments](https://github.com/videojs/video.js/pull/1833)

Comment (heff) [inline src/js/lib.js]: Ah, right. Thanks.

File: src/js/media/flash.js (hunk 2 above)
Comment (carpasse) [inline src/js/media/flash.js]: I believe it clarifies the intention. I can remove it if you don't like it.

File: test/unit/lib.js (hunk 3 above)
Comment (carpasse) [inline test/unit/lib.js]: I have added some tests cases with 'http://' is that enough??

File: src/js/media/flash.js (hunk 2 above)
Comment (heff) [inline src/js/media/flash.js]: Cool, I can see that. Works for me.

File: test/unit/lib.js (hunk 3 above)
Comment (heff) [inline test/unit/lib.js]: I think that's good. If anyone else has ideas on URL formats we should be testing for speak up.

**Comment by heff** [comment]
I think this is ready to go. Just need one more person to review and sign off on it. @videojs/core-committers

**Comment by xy2015** [comment]
Sent from Windows Mail





From: Steve Heffernan
Sent: ‎Wednesday‎, ‎March‎ ‎25‎, ‎2015 ‎7‎:‎25‎ ‎AM
To: videojs/video.js





Added a few comments, but otherwise this looks good to me.  :+1:

**Comment by carpasse** [comment]
@xy2015 I can not see you comments can you please add them again or point me to them??

File: test/unit/lib.js (hunk 3 above)
Comment (dmlap) [inline test/unit/lib.js]: A protocol-relative URL (e.g. "//example.com/movie.mp4") would be nice but I don't think it needs to hold up merging this.

**Comment by dmlap** [comment]
LGTM

**Comment by dmlap** [comment]
@heff I hit the "merge" button by habit, sorry. I'll fix up the changelog tomorrow morning

**Comment by heff** [comment]
No problem. Though @mmcc might not have fun rebasing the ES6 work again. :smile:

**Comment by mmcc** [comment]
(╯°□°）╯︵ ┻━┻ 




—

**Comment by dmlap** [comment]
@mmcc sorry!

**Comment by mmcc** [comment]
Ok this is now pulled into ES6 branch, but I'm going to hold off on pushes until all the comments are addressed. I've rebased from master to clean up the commit history so I think we'll lose all the comments when I end up force pushing.

**Pull request merged**